src/backend/distributed/utils/placement_connection.c

```diff
@@ -0,0 +1,574 @@
+	 * establishment.  Allocations are performed in transaction context, so we
+	 * don't have to care about freeing in case of an early disconnect.
+	 */
+	if (returnConnectionReference == NULL)
+	{
+		MultiConnection *connection = StartNodeConnection(flags, placement->nodeName,
+														  placement->nodePort);
+
+		returnConnectionReference = (ConnectionReference *)
+									MemoryContextAlloc(TopTransactionContext,
+													   sizeof(ConnectionReference));
+		returnConnectionReference->connection = connection;
+		returnConnectionReference->hadDDL = false;
+		returnConnectionReference->hadDML = false;
+		returnConnectionReference->shardId = placement->shardId;
+		returnConnectionReference->placementId = placement->placementId;
+		returnConnectionReference->userName =
+			MemoryContextStrdup(TopTransactionContext, userName);
+		dlist_push_tail(&placementEntry->connectionReferences,
+						&returnConnectionReference->placementNode);
+
+		/* record association with shard, for invalidation */
+		AssociatePlacementWithShard(placementEntry, placement);
+
+		/* record association with connection, to handle connection closure */
+		dlist_push_tail(&connection->referencedPlacements,
+						&returnConnectionReference->connectionNode);
+	}
+
+	if (flags & FOR_DDL)
+	{
+		returnConnectionReference->hadDDL = true;
+	}
+	if (flags & FOR_DML)
+	{
+		returnConnectionReference->hadDML = true;
+	}
+
+	if (freeUserName)
+	{
+		pfree(freeUserName);
+	}
+
+	return returnConnectionReference->connection;
+}
+
+
+/*
+ * CheckExistingConnections check whether any of the existing connections is
+ * usable. If so, return it, otherwise return NULL.
+ *
+ * A connection is usable if it is not in use, the user matches, DDL/DML usage
+ * match and cached connection are allowed.  If no existing connection
+ * matches, but a new connection would conflict (e.g. a connection already
+ * exists but isn't usable, and the desired connection needs to execute
+ * DML/DML) an error is thrown.
+ */
+static ConnectionReference *
+CheckExistingConnections(uint32 flags, const char *userName,
+						 ConnectionPlacementHashEntry *placementEntry)
+{
+	dlist_iter it;
+	ConnectionReference *returnConnectionReference = NULL;
+
+	dlist_foreach(it, &placementEntry->connectionReferences)
+	{
+		ConnectionReference *connectionReference =
+			dlist_container(ConnectionReference, placementNode, it.cur);
+		MultiConnection *connection = connectionReference->connection;
+		bool useConnection = false;
+
+		if (!connection)
+		{
+			/* if already closed connection obviously not usable */
+			useConnection = false;
+		}
+		else if (connection->claimedExclusively)
+		{
+			/* already used */
+			useConnection = false;
+		}
+		else if (flags & FORCE_NEW_CONNECTION)
+		{
+			/* no connection reuse, just check for conflicts */
+			useConnection = false;
+		}
+		else if (returnConnectionReference != NULL)
+		{
+			/* already found connection, just check for conflicts */
+			useConnection = false;
```

Comment:
Could there be a situation in which the list of connections contains 2 usable connections and the first one only read from the placement while the second one has returned DML/DDL, and the current connection is FOR_DML/FOR_DDL? Hence we error out instead of returning the modifying connection?